cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java

```diff
@@ -274,16 +279,42 @@ private Set<Range<BigInteger>> getIntersectingSubRanges(Set<Range<BigInteger>> r
                      .collect(Collectors.toSet());
     }
 
-    private boolean haveTokenRangeMappingsChanged(TokenRangeMapping<RingInstance> startTaskMapping, Range<BigInteger> taskTokenRange)
+    private void validateTaskTokenRangeMappings(int partitionId,
+                                                TokenRangeMapping<RingInstance> startTaskMapping,
+                                                Range<BigInteger> taskTokenRange)
     {
         // Get the uncached, current view of the ring to compare with initial ring
         TokenRangeMapping<RingInstance> endTaskMapping = writerContext.cluster().getTokenRangeMapping(false);
         Map<Range<BigInteger>, List<RingInstance>> startMapping = taskTokenRangeMapping(startTaskMapping, taskTokenRange);
         Map<Range<BigInteger>, List<RingInstance>> endMapping = taskTokenRangeMapping(endTaskMapping, taskTokenRange);
 
+        Set<RingInstance> initialInstances = instancesFromMapping(startMapping);
+        Set<RingInstance> endInstances = instancesFromMapping(endMapping);
         // Token ranges are identical and overall instance list is same
-        return !(startMapping.keySet().equals(endMapping.keySet()) &&
-               instancesFromMapping(startMapping).equals(instancesFromMapping(endMapping)));
+        boolean haveMappingsChanged = !(startMapping.keySet().equals(endMapping.keySet()) &&
+                                        initialInstances.equals(endInstances));
+        if (haveMappingsChanged)
+        {
+            Set<Range<BigInteger>> rangeDelta = symmetricDifference(startMapping.keySet(), endMapping.keySet());
+            Set<String> instanceDelta = symmetricDifference(initialInstances, endInstances).stream()
+                                                                                           .map(RingInstance::getIpAddress)
+                                                                                           .collect(Collectors.toSet());
+            String message = String.format("[%s] Token range mappings have changed since the task started " +
+                                           "with non-overlapping instances: %s and ranges: %s",
+                                           partitionId,
+                                           instanceDelta,
+                                           rangeDelta);
+            LOGGER.error(message);
+            throw new RuntimeException(message);
+        }
+    }
+
+    public static <T> Set<T> symmetricDifference(Set<T> set1, Set<T> set2)
```

Comment:
if feels we should be able to test this code. Maybe add some unit tests for this?

Reply:
Will do